File: 20-actors/mitooshi/methods/test_promote.py
#!/usr/bin/env python3
"""Tests for mitooshi backtest→promotion decision (methods/promote.py).

    PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 python3 -m pytest test_promote.py
    python3 test_promote.py

The calibration_gate is a REFUSAL gate. These tests prove each gate fires on the scorecard:
G12 (skill≤0 refused), G7 (miscalibrated refused — the real two-regime trail's outcome),
G9 (unsigned / server-signed refused), and that a skilled+calibrated+member-signed model
clears. The gate logic itself is the cell's review_promotion (single source of truth).
"""
from __future__ import annotations

import pathlib
import sys

try:
    from promote import decide_from_scorecard, emit_decision_edn
    from analyze import load_edn
except ImportError:
    from mitooshi.methods.promote import decide_from_scorecard, emit_decision_edn  # type: ignore
    from mitooshi.methods.analyze import load_edn  # type: ignore

_SCORECARD = (pathlib.Path(__file__).resolve().parent.parent / "data" / "persisted"
              / "chokepoint-backtest-scorecard.kotoba.edn")
_MEMBER = "did:web:etzhayyim.com:member:alice"


def _rows(skill, deviation, method="persistence"):
    return [{":fc.score/method": f":{method}", ":fc.score/mean-skill": skill,
             ":fc.score/calibration-deviation": deviation}]


def test_skilled_calibrated_signed_clears():
    d = decide_from_scorecard(_rows(0.5, 0.2), signed_by=_MEMBER)[0]
    assert d["phase"] == "cleared" and d["promoted"] is True


def test_unskilled_refused_g12():
    d = decide_from_scorecard(_rows(-0.1, 0.2), signed_by=_MEMBER)[0]
    assert d["phase"] == "refused" and "G12" in d["refusal"]


def test_miscalibrated_refused_g7():
    # skill is fine but deviation exceeds the ceiling → G7 refusal
    d = decide_from_scorecard(_rows(0.7, 1.3), signed_by=_MEMBER)[0]
    assert d["phase"] == "refused" and "G7" in d["refusal"]


def test_unsigned_refused_g9_no_server_key():
    d = decide_from_scorecard(_rows(0.5, 0.2), signed_by="")[0]
    assert d["phase"] == "refused" and "G9" in d["refusal"]


def test_server_signature_refused_g9():
    d = decide_from_scorecard(_rows(0.5, 0.2), signed_by="server:etzhayyim")[0]
    assert d["phase"] == "refused" and "G9" in d["refusal"]


def test_decision_edn_records_server_held_key_false():
    edn = emit_decision_edn(decide_from_scorecard(_rows(0.5, 0.2), signed_by=_MEMBER), _MEMBER)
    assert ":fc.promotion/server-held-key false" in edn
    assert ":fc.promotion/promoted true" in edn


def test_real_scorecard_is_refused_on_calibration():
    # honest end-to-end: the real two-regime trail is SKILLED but MISCALIBRATED, so even
    # a member signature does NOT clear it — the gate working as designed.
    if not _SCORECARD.exists():
        return
    rows = load_edn(_SCORECARD)
    decisions = decide_from_scorecard(rows, signed_by=_MEMBER)
    assert decisions, "expected scorecard methods"
    for d in decisions:
        assert d["skill"] > 0                      # G12 satisfied (skilled)
        assert d["phase"] == "refused" and "G7" in d["refusal"]   # but miscalibrated


def _run():
    fns = [v for k, v in sorted(globals().items())
           if k.startswith("test_") and callable(v)]
    for fn in fns:
        fn()
    print(f"promote.py: {len(fns)}/{len(fns)} tests passed")
    return True


if __name__ == "__main__":
    sys.exit(0 if _run() else 1)

File: 20-actors/mitooshi/methods/test_promote.clj
#!/usr/bin/env bb
;; Tests for mitooshi backtest→promotion decision (methods/promote.clj).
;; 1:1 port of test_promote.py — all 7 test cases, every assertion preserved.
;;
;; Run:
;;   bb --classpath 20-actors 20-actors/mitooshi/methods/test_promote.clj
;;
;; The .cljc sibling (test_promote.cljc) defines the same ns but its -main is never
;; invoked by bb (hollow false-green). This .clj file runs exactly the 7 tests and
;; confirms "Ran 7 tests" — proof the correct file loaded.
(ns mitooshi.methods.test-promote
  "Tests for mitooshi backtest→promotion decision (methods/promote.clj).
  1:1 port of methods/test_promote.py.

  The calibration_gate is a REFUSAL gate. These tests prove each gate fires on the
  scorecard: G12 (skill≤0 refused), G7 (miscalibrated refused — the real two-regime
  trail's outcome), G9 (unsigned / server-signed refused), and that a
  skilled+calibrated+member-signed model clears. The gate logic itself is the cell's
  review-promotion (single source of truth)."
  (:require [clojure.test :refer [deftest is run-tests]]
            [clojure.string :as str]
            [clojure.java.io :as io]
            [mitooshi.methods.promote :as p]))

;; PROOF this .clj loaded (not the .cljc): resolve a var present only in promote.clj.
;; If bb loaded the .cljc, this would throw and tests would not run.
(when-not (resolve 'mitooshi.methods.promote/promote-clj-loaded)
  (throw (ex-info "bb loaded promote.cljc instead of promote.clj — wrong file!" {})))

(def ^:private member "did:web:etzhayyim.com:member:alice")

;; The scorecard the Python test reads: methods/../data/persisted/...
;; *file* is the path of this test file: .../methods/test_promote.clj
(def ^:private scorecard-file
  (-> (io/file *file*)        ;; .../methods/test_promote.clj
      (.getParentFile)        ;; .../methods
      (.getParentFile)        ;; .../mitooshi
      (io/file "data" "persisted" "chokepoint-backtest-scorecard.kotoba.edn")))

(defn- rows
  "Build a minimal scorecard row vector — mirrors _rows() in test_promote.py."
  ([skill deviation] (rows skill deviation "persistence"))
  ([skill deviation method]
   [{":fc.score/method" (str ":" method)
     ":fc.score/mean-skill" skill
     ":fc.score/calibration-deviation" deviation}]))

;; ── test cases (1:1 with test_promote.py) ────────────────────────────────────

(deftest test-skilled-calibrated-signed-clears
  ;; skilled (G12 ok), calibrated (G7 ok), member-signed (G9 ok) → CLEARED
  (let [d (first (p/decide-from-scorecard (rows 0.5 0.2) {:signed-by member}))]
    (is (= "cleared" (:phase d)))
    (is (true? (:promoted d)))))

(deftest test-unskilled-refused-g12
  ;; skill ≤ 0 → G12 refusal
  (let [d (first (p/decide-from-scorecard (rows -0.1 0.2) {:signed-by member}))]
    (is (= "refused" (:phase d)))
    (is (str/includes? (:refusal d) "G12"))))

(deftest test-miscalibrated-refused-g7
  ;; skill is fine but deviation exceeds the ceiling → G7 refusal
  (let [d (first (p/decide-from-scorecard (rows 0.7 1.3) {:signed-by member}))]
    (is (= "refused" (:phase d)))
    (is (str/includes? (:refusal d) "G7"))))

(deftest test-unsigned-refused-g9-no-server-key
  ;; no signed-by → G9 refusal
  (let [d (first (p/decide-from-scorecard (rows 0.5 0.2) {:signed-by ""}))]
    (is (= "refused" (:phase d)))
    (is (str/includes? (:refusal d) "G9"))))

(deftest test-server-signature-refused-g9
  ;; signed-by starting with "server" → G9 refusal (no-server-key)
  (let [d (first (p/decide-from-scorecard (rows 0.5 0.2) {:signed-by "server:etzhayyim"}))]
    (is (= "refused" (:phase d)))
    (is (str/includes? (:refusal d) "G9"))))

(deftest test-decision-edn-records-server-held-key-false
  ;; emit-decision-edn always records serverHeldKey false (no-server-key) + promoted true
  (let [edn (p/emit-decision-edn
             (p/decide-from-scorecard (rows 0.5 0.2) {:signed-by member})
             member)]
    (is (str/includes? edn ":fc.promotion/server-held-key false"))
    (is (str/includes? edn ":fc.promotion/promoted true"))))

(deftest test-real-scorecard-is-refused-on-calibration
  ;; honest end-to-end: the real two-regime trail is SKILLED but MISCALIBRATED, so even
  ;; a member signature does NOT clear it — the gate working as designed, not a bug.
  (when (.exists scorecard-file)
    (let [rows      (p/load-edn scorecard-file)
          decisions (p/decide-from-scorecard rows {:signed-by member})]
      (is (seq decisions) "expected scorecard methods")
      (doseq [d decisions]
        (is (> (:skill d) 0))                          ;; G12 satisfied (skilled)
        (is (= "refused" (:phase d)))
        (is (str/includes? (:refusal d) "G7"))))))     ;; but miscalibrated → G7 refused

;; ── entry point ──────────────────────────────────────────────────────────────

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (clojure.test/run-tests 'mitooshi.methods.test-promote)]
    (System/exit (if (zero? (+ fail error)) 0 1))))
